server/service/src/invoice/inventory_adjustment/insert.rs

```diff
@@ -0,0 +1,245 @@
+        //     Err(ServiceError::AdjustmentReasonNotProvided)
+        // );
+
+        // Invalid reason
+        assert_eq!(
+            service.insert_inventory_adjustment(
+                &context,
+                InsertInventoryAdjustment {
+                    stock_line_id: mock_stock_line_a().id,
+                    inventory_adjustment_reason_id: Some("invalid".to_string()),
+                    ..Default::default()
+                }
+            ),
+            Err(ServiceError::InvalidAdjustment)
+        );
+
+        // Reduce stock below zero
+        assert_eq!(
+            service.insert_inventory_adjustment(
+                &context,
+                InsertInventoryAdjustment {
+                    stock_line_id: mock_stock_line_a().id,
+                    adjustment: 0.0,
+                    ..Default::default()
+                }
+            ),
+            Err(ServiceError::InvalidAdjustment)
+        );
+
+        // Reduce stock below zero
+        // assert_eq!(
+        //     service.insert_inventory_adjustment(
+        //         &context,
+        //         InsertInventoryAdjustment {
+        //             stock_line_id: mock_stock_line_a().id,
+        //             adjustment_type:
+        //                 crate::invoice::inventory_adjustment::AdjustmentType::Reduction,
+        //             adjustment: 50.0,
+        //             ..Default::default()
+        //         }
+        //     ),
+        //     Err(ServiceError::StockLineReducedBelowZero(StockLine {
+        //         stock_line_row: mock_stock_line_a(),
+        //         ..Default::default()
+        //     }))
+        // );
+    }
+
+    #[actix_rt::test]
+    async fn insert_inventory_adjustment_success() {
+        let (_, connection, connection_manager, _) =
+            setup_all("insert_inventory_adjustment_errors", MockDataInserts::all()).await;
+
+        let service_provider = ServiceProvider::new(connection_manager, "app_data");
+        let context = service_provider
+            .context(mock_store_a().id, mock_user_account_a().id)
+            .unwrap();
+        let service = service_provider.invoice_service;
+
+        // Success
+        let created_invoice = service
+            .insert_inventory_adjustment(
+                &context,
+                InsertInventoryAdjustment {
+                    stock_line_id: mock_stock_line_a().id,
+                    adjustment: 2.0,
+                    ..Default::default()
+                },
+            )
+            .unwrap();
+
+        let retrieved_invoice = InvoiceRowRepository::new(&connection)
+            .find_one_by_id(&created_invoice.invoice_row.id)
+            .unwrap();
+
+        assert_eq!(
+            retrieved_invoice,
+            inline_edit(&retrieved_invoice, |mut u| {
+                u.id = created_invoice.invoice_row.id;
+                u.status = InvoiceRowStatus::Verified;
+                u
+            })
+        );
```

Comment:
Should we also retrieve the new invoice line and check it? 

What other types of "Success" tests should we have?

Reply:
Yeah could check the invoice line looks as expected. Probably more important though would be to query the stock line, and check the stock levels have been adjusted accordingly.

So I would write 2 success tests, one for addition and one for reduction 👍

Reply:
Probably also good to have a quick check in one of the success tests that if no reasons configured, the reason not provided error _doesn't_ throw